test/cpp/end2end/tls_private_key_offload_end2end_test.cc

```diff
@@ -0,0 +1,422 @@
+    default:
+      return false;
+  }
+}
+
+class TestCustomPrivateKeySigner final
+    : public grpc_core::CustomPrivateKeySigner {
+ public:
+  explicit TestCustomPrivateKeySigner(absl::string_view private_key)
+      : pkey_(LoadPrivateKeyFromString(private_key)) {}
+
+  void Sign(absl::string_view data_to_sign,
+            SignatureAlgorithm signature_algorithm,
+            OnSignComplete on_sign_complete) override {
+    auto on_sign_complete_ptr =
+        std::make_shared<OnSignComplete>(std::move(on_sign_complete));
+    LOG(ERROR) << "TestCustomPrivateKeySigner schedule length "
+               << data_to_sign.length();
+    std::thread([this, data_to_sign = std::string(data_to_sign),
+                 signature_algorithm, on_sign_complete_ptr]() mutable {
+      const EVP_MD* md = nullptr;
+      int padding = 0;
+      if (!GetBoringSslAlgorithm(signature_algorithm, &md, &padding)) {
+        (*on_sign_complete_ptr)(
+            absl::InternalError("Unsupported signature algorithm"));
+        return;
+      }
+      bssl::ScopedEVP_MD_CTX ctx;
+      EVP_PKEY_CTX* pctx = nullptr;
+      if (!EVP_DigestSignInit(ctx.get(), &pctx, md, nullptr, pkey_.get())) {
+        (*on_sign_complete_ptr)(
+            absl::InternalError("EVP_DigestSignInit failed"));
+        return;
+      }
+      if (padding == RSA_PKCS1_PADDING) {
+        if (!EVP_PKEY_CTX_set_rsa_padding(pctx, RSA_PKCS1_PADDING)) {
+          (*on_sign_complete_ptr)(
+              absl::InternalError("EVP_PKEY_CTX_set_rsa_padding failed"));
+          return;
+        }
+      } else if (padding == RSA_PKCS1_PSS_PADDING) {
+        if (!EVP_PKEY_CTX_set_rsa_padding(pctx, RSA_PKCS1_PSS_PADDING) ||
+            !EVP_PKEY_CTX_set_rsa_pss_saltlen(pctx, -1)) {
+          (*on_sign_complete_ptr)(
+              absl::InternalError("EVP_PKEY_CTX_set_rsa_padding failed"));
+          return;
+        }
+      }
+      size_t sig_len = 0;
+      if (!EVP_DigestSignUpdate(ctx.get(), data_to_sign.data(),
+                                data_to_sign.size()) ||
+          !EVP_DigestSignFinal(ctx.get(), nullptr, &sig_len)) {
+        (*on_sign_complete_ptr)(
+            absl::InternalError("EVP_DigestSignFinal failed"));
+        return;
+      }
+      std::string sig;
+      sig.resize(sig_len);
+      if (!EVP_DigestSignFinal(
+              ctx.get(), reinterpret_cast<uint8_t*>(sig.data()), &sig_len)) {
+        (*on_sign_complete_ptr)(
+            absl::InternalError("EVP_DigestSignFinal failed"));
+        return;
+      }
+      sig.resize(sig_len);
+      LOG(ERROR) << "TestCustomPrivateKeySigner result length " << sig_len;
+      (*on_sign_complete_ptr)(std::move(sig));
+    }).detach();
+  }
+
+  ~TestCustomPrivateKeySigner() {}
+
+ private:
+  bssl::UniquePtr<EVP_PKEY> pkey_;
+};
+
+TEST_F(TlsPrivateKeyOffloadTest, DefaultOffload) {
+  server_addr_ = absl::StrCat("localhost:",
+                              std::to_string(grpc_pick_unused_port_or_die()));
```

Comment:
No need for `std::to_string()` here, since `absl::StrCat()` does the same thing automatically.

Same thing throughout.

Reply:
Doesn't look like this was addressed.